airbyte-integrations/connectors/source-shopify/source_shopify/source.py

```diff
@@ -0,0 +1,226 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+# import time
+import math
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+import requests
+from airbyte_cdk import AirbyteLogger
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.models import SyncMode
+
+
+class ShopifyStream(HttpStream, ABC):
+    url_base = ""
+    primary_key = "id"
+    limit = 200
+
+    def __init__(self, shop: str, api_key: str, api_password: str, api_version: str, **kwargs):
+        super().__init__(**kwargs)
+        self.shop = shop
+        self.api_key = api_key
+        self.api_password = api_password
+        self.api_version = api_version
+        self.since_id = 0
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        decoded_response = response.json()
+        if len(decoded_response.get(self.data_field)) < self.limit:
+            return None
+        else:
+            self.since_id = decoded_response.get(self.data_field)[-1]["id"]
+            # UNCOMMENT FOR DEBUG ONLY
+            # print(f"\nSINCE_ID TOKEN: {self.since_id}\n")
+            # time.sleep(5)
+            return {"since_id": self.since_id}
+
+    def request_params(
+        self,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+        **kwargs
+        ) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        params["limit"] = self.limit
+        params["since_id"] = self.since_id
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        json_response = response.json()
+        records = json_response.get(self.data_field, []) if self.data_field is not None else json_response
+        for record in records:
+            yield record
+
+    @property
+    @abstractmethod
+    def data_field(self) -> str:
+        """The name of the field in the response which contains the data"""
+
+
+# Basic incremental stream
+class IncrementalShopifyStream(ShopifyStream, ABC):
```

Comment:
This incremental implementation looks sound! small comment but otherwise I think it LGTM